### PR TITLE
Verilator v5

### PR DESCRIPTION
**Pull request by GregAC**
I'd like this working for the hackathon so if you've got more major changes you'd like following review I'd appreciate it if we could just record it in an issue for now and get this merged.

### CODE REVIEW

**Comment by HU90m** [comment]
This looks brilliant, thanks Greg!

I'll quickly change the GitHub actions to do simulator builds in the nix sandbox, so we don't have to keep updating the dependencies.

**Comment by GregAC** [comment]
> This looks brilliant, thanks Greg!
> 
> I'll quickly change the GitHub actions to do simulator builds in the nix sandbox, so we don't have to keep updating the dependencies.

Thanks I've just added a commit to this PR to do just that but as you know nix better than me let's see what you propose as well!

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -50,12 +50,14 @@ jobs:
     steps:
     - uses: actions/checkout@v4
 
-    - name: Install dependencies
-      run: |
-        sudo apt install verilator libelf-dev
-        pip install -r python-requirements.txt
+    - name: Install Nix
+      uses: cachix/install-nix-action@v24
```
Comment (HU90m) [inline .github/workflows/ci.yml:54]:
```suggestion
      uses: cachix/install-nix-action@v24
      
    - name: Nix Format Check
      run: nix fmt -- . --check
```

May as well add a quick Nix format check while your at it.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -50,12 +50,14 @@ jobs:
...
     - name: Build Verilator simulator
       run: |
-        fusesoc --cores-root=. run --target=sim --tool=verilator --setup --build lowrisc:sonata:system --verilator_options="+define+RVFI"
-
+        # TODO: Actually build with nix so this tests the sonata-simulator
+        # package build
+        nix develop . -c fusesoc --cores-root=. run --target=sim \
+          --tool=verilator --setup --build lowrisc:sonata:system \
+          --verilator_options="+define+RVFI -j $NIX_BUILD_CORES" \
+          --make_options="-j $NIX_BUILD_CORES"
```
Comment (HU90m) [inline .github/workflows/ci.yml:63]:
```suggestion
        nix build .#sonata-simulator
```

This'll build the simulator in a sandbox

**Review by HU90m — CHANGES_REQUESTED** [review]

**Comment by GregAC** [comment]
Thanks @HU90m, any thoughts on how to make it use the lowRISC cache? It spends the first few minutes building python dependencies

**Comment by HU90m** [comment]
This should work for now:

```sh
nix build --accept-flake-config .#sonata-simulator
```

It would be nice to get the `sonata-system` CI pushing to the cache and running on pomona, espacially once https://github.com/lowRISC/sonata-system/pull/98 is merged. @nbdd0121 (tagging for visibility, no immediate action required.)

File: sw/cheri/sim_boot_stub/makefile
Comment (nbdd0121) [inline sw/cheri/sim_boot_stub/makefile:1]: `Makefile` per convention

File: sw/cheri/sim_boot_stub/makefile [hunk 3]
```diff
@@ -0,0 +1,9 @@
+CFLAGS=-target riscv32-unknown-unknown -mcpu=cheriot -mabi=cheriot \
```
Comment (nbdd0121) [inline sw/cheri/sim_boot_stub/makefile:1]: It'll be good that this uses cmake similar to other binaries that we have, but we can save that to a later PR.

File: flake.nix [hunk 4]
```diff
@@ -36,23 +36,28 @@
...
-          [lrPkgs.python_ot];
+          [lrPkgs.python_ot] ++ cheriotPkgs ;
         buildPhase = ''
           HOME=$TMPDIR fusesoc --cores-root=. run \
             --target=sim --tool=verilator --setup \
             --build lowrisc:sonata:system \
             --verilator_options="+define+RVFI -j $NIX_BUILD_CORES" \
-            --make_options="-j $NIX_BUILD_CORES"
+            --make_options="-j $NIX_BUILD_CORES" &&
+            make -C sw/cheri/sim_boot_stub/
```
Comment (nbdd0121) [inline flake.nix:56]: We probably should put the sim stub building in another derivation.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -50,12 +50,12 @@ jobs:
     steps:
     - uses: actions/checkout@v4
 
-    - name: Install dependencies
-      run: |
-        sudo apt install verilator libelf-dev
-        pip install -r python-requirements.txt
+    - name: Install Nix
+      uses: cachix/install-nix-action@v24
```
Comment (nbdd0121) [inline .github/workflows/ci.yml:54]: To enable cache globally for all nix commands, you can do something similar to https://github.com/lowRISC/lowrisc-nix/blob/34f1d472d6066a42c684416f5e9e755d25ec6270/.github/workflows/ci.yml#L52-L54 here.

File: flake.nix [hunk 6]
```diff
@@ -43,8 +43,10 @@
         pname = "sonata-simulator";
         src = ./.;
         buildInputs = with pkgs; [libelf zlib];
-        nativeBuildInputs = (with pkgs; [python311Packages.pip verilator]) ++
-          [lrPkgs.python_ot] ++ cheriotPkgs ;
+        nativeBuildInputs =
```
Comment (nbdd0121) [inline flake.nix:46]: Squash this commit?

File: flake.nix (hunk 4 above)
Comment (HU90m) [reply flake.nix:56]: I had a similar thought, but I'm happy for this to go in for now. In a follow up PR, we can transition to CMake and create derivations for the software builds.

File: sw/cheri/sim_boot_stub/makefile (hunk 3 above)
Comment (GregAC) [reply sw/cheri/sim_boot_stub/makefile:1]: It uses make intentionally so you can build it in environments where you don't have cmake just to minimize dependencies required to build it

File: flake.nix (hunk 4 above)
Comment (GregAC) [reply flake.nix:56]: I did actually do it that way initially, then switched it over to this as it felt like it's something you should distribute with the simulator

**Review by marnovandermaas — APPROVED** [review]
Works for me. Happy to merge this once CI passes.

File: flake.nix (hunk 4 above)
Comment (nbdd0121) [reply flake.nix:69]: What could be done is to build it in a separate derivation, and then copy it to $out/share for the simulator.

Comment (GregAC) [reply flake.nix:69]: Turns out we need to do this to get the osx build to work. CHERIoT LLVM clashes with the system LLVM and the verilated C++ fails to build so will make that change.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (marnovandermaas) [reply .github/workflows/ci.yml]: resolved, I believe.

**Comment by GregAC** [comment]
@HU90m happy to approve this now?

**Review by nbdd0121 — APPROVED** [review]

**Review by HU90m — APPROVED** [review]

**Comment by marnovandermaas** [comment]
You force pushed over my documentation change, I've made a follow up PR here: https://github.com/lowRISC/sonata-system/pull/105

**Pull request merged**